File: src/main/java/net/moltendorf/Bukkit/SkyrimAppetite/Listeners.java
package net.moltendorf.Bukkit.SkyrimAppetite;

import org.bukkit.Material;
import org.bukkit.block.Block;
import org.bukkit.entity.Player;
import org.bukkit.event.EventHandler;
import org.bukkit.event.EventPriority;
import org.bukkit.event.Listener;
import org.bukkit.event.player.PlayerInteractEvent;
import org.bukkit.event.player.PlayerItemConsumeEvent;
import org.bukkit.inventory.ItemStack;
import org.bukkit.material.MaterialData;

import java.util.HashMap;
import java.util.Map;
import java.util.UUID;

/**
 * Created by moltendorf on 2016-02-28.
 */
public class Listeners implements Listener {
	final protected Map<UUID, Integer> foodLevels = new HashMap<>();

	private void playerConsumedFood(Player player, Material material, MaterialData data) {

		UUID playerId = player.getUniqueId();

		Integer foodLevel = foodLevels.get(playerId);

		if (foodLevel == null) {
			foodLevel = 0;

			player.sendMessage("Since this is your first time eating food with SkyrimAppetite installed, here's how hunger works: we count how much food you've eaten; if you've eaten enough, you gain a point of hunger.");
			player.sendMessage("In short, you must eat a ton of food now.");
		}

		Settings settings = Settings.getInstance();

		Integer foodValue = settings.getFoodValue(material, data);
		foodLevel += foodValue;

		int maxFoodLevel = settings.getMaxFoodLevel();

		if (foodLevel >= maxFoodLevel) {
			player.setFoodLevel(player.getFoodLevel() - foodValue + 1);

			foodLevel -= maxFoodLevel;
		} else {
			player.setFoodLevel(player.getFoodLevel() - foodValue);
		}

		foodLevels.put(playerId, foodLevel);
	}

	@EventHandler(priority = EventPriority.MONITOR)
	public void playerItemConsumeEventMonitor(PlayerItemConsumeEvent event) {
		ItemStack item = event.getItem();

		playerConsumedFood(event.getPlayer(), item.getType(), item.getData());
	}

	@EventHandler(priority = EventPriority.MONITOR)
	public void playerInteractEventMonitor(PlayerInteractEvent event) {
		Block block = event.getClickedBlock();

		if (block != null) {
			Material material = block.getType();
			Player player = event.getPlayer();

			if (material == Material.CAKE_BLOCK && player.getFoodLevel() < 20) {
				playerConsumedFood(player, material, material.getNewData((byte) 0));
			}
		}
	}
}

File: src/main/kotlin/net/moltendorf/bukkit/skyrimappetite/Listeners.kt
package net.moltendorf.bukkit.skyrimappetite

import org.bukkit.Material
import org.bukkit.entity.Player
import org.bukkit.event.EventHandler
import org.bukkit.event.EventPriority
import org.bukkit.event.Listener
import org.bukkit.event.player.PlayerInteractEvent
import org.bukkit.event.player.PlayerItemConsumeEvent
import org.bukkit.material.MaterialData

/**
 * Created by moltendorf on 2016-02-28.
 */
class Listeners : Listener {
  private fun playerConsumedFood(player: Player, material: Material, data: MaterialData) {
    val playerId = player.uniqueId
    var foodLevel: Int? = settings.foodLevels[playerId]
    val foodValue = settings.getFoodValue(material, data)
    val playerFoodLevel = player.foodLevel
    val statusMultiplier = if (playerFoodLevel >= settings.foodMidpoint) settings.foodMidpoint / 10.0 else 10.0 / settings.foodMidpoint
    val foodMultiplier = settings.foodMultiplier * statusMultiplier

    if (foodLevel == null) {
      foodLevel = 0

      player.sendMessage("Since this is your first time eating food with SkyrimAppetite installed, here's how hunger works: we count how much food you've eaten; if you've eaten enough, you gain a point of hunger.")
      player.sendMessage("In short, you must eat a ton of food now.")
    }

    foodLevel += foodValue

    if (foodLevel >= foodMultiplier || playerFoodLevel <= 0) {
      val foodPoints = (foodLevel / foodMultiplier).toInt()

      player.foodLevel = playerFoodLevel - foodValue + foodPoints

      foodLevel = (foodLevel % foodMultiplier).toInt()
    } else {
      player.foodLevel = playerFoodLevel - foodValue
    }

    settings.foodLevels[playerId] = foodLevel
  }

  @EventHandler(priority = EventPriority.MONITOR)
  fun playerItemConsumeEventMonitor(event: PlayerItemConsumeEvent) {
    val item = event.item

    playerConsumedFood(event.player, item.type, item.data)
  }

  @EventHandler(priority = EventPriority.MONITOR)
  fun playerInteractEventMonitor(event: PlayerInteractEvent) {
    val block = event.clickedBlock ?: return
    val material = block.type
    val player = event.player

    if (material == Material.CAKE_BLOCK && player.foodLevel < 20) {
      playerConsumedFood(player, material, material.getNewData(0.toByte()))
    }
  }
}
